### PR TITLE
[Development branch] Generate AppImage for each git push

### PR DESCRIPTION
**Pull request by probonopd**
screenshot from 2017-04-15 17-02-00

Closes https://github.com/gridcoin/Gridcoin-Research/issues/278 and https://github.com/gridcoin/Gridcoin-Research/issues/280

This PR, when merged, will compile this application on [Travis CI](https://travis-ci.org/) upon each `git push`, and upload an [AppImage](http://appimage.org/) to a temporary download URL on transfer.sh (available for 14 days). The download URL is toward the end of each Travis CI build log of each build (see below for how to set up automatic uploading to your GitHub Releases page).

For this to work, you need to enable Travis CI for your repository as [described here](https://travis-ci.org/getting_started) __prior to merging this__, if you haven't already done so.

__Please note:__ Instead of storing AppImage builds temporarily for 14 days each on transfer.sh, you could use GitHub Releases to store the binaries permanently. This way, they would be visible on the Releases page of your project. This is what I recommend. See https://docs.travis-ci.com/user/deployment/releases/. If you want to do this for continuous builds, also see https://github.com/probonopd/uploadtool.

If you would like to see only one entry for the Pull Request in your project's history, then please enable [this GitHub functionality](https://help.github.com/articles/configuring-commit-squashing-for-pull-requests/) on your repo. It allows you to squash (combine) the commits when merging.

But I get:

> To get started with Boinc, run the boinc client, choose projects, then populate the gridcoinresearch.conf file in %appdata%\GridcoinResearch with your boinc e-mail address.  To run this wizard again, please delete the gridcoinresearch.conf file. 
> 
> Boinc is not installed in default location /var/lib/boinc-client/!  Please set boincdatadir=c:\programdata\boinc\    to the correct path where Boincs programdata directory resides.

Seems like we should bundle it inside the AppImage?

If you have questions, AppImage developers are on #AppImage on irc.freenode.net.

### CODE REVIEW

**Comment by philipswift** [comment]
Has the path changed to single backslashes for Win? Above mentions it as boincdatadir=c:\programdata\boinc\ (single backslash)
**Data**
It was boincdatadir=c:\\\programdata\\\boinc\\\\ (double backslash)
Linux was boincdatadir=/var/lib/boinc-client/

**Application**
The same was for application location under Win for non-default installation EG:
boincappdir=d:\\\Program files\\\boinc\\\

**Comment by probonopd** [comment]
I have not changed this. So this must be unrelated to this PR @philipswift but indeed it looks like Windows strings are displayed in this Linux build...

**Comment by denravonska** [comment]
Pretty damn sweet. I believe Travis is enabled for this repo but not activated. Do we need to wait until activation?

**Comment by probonopd** [comment]
@denravonska looks like this repository needs to be enabled in Travis CI:
https://travis-ci.org/gridcoin/Gridcoin-Research/

https://travis-ci.org/getting_started

**Comment by Peppernrino** [comment]
same problem as in this thread: https://github.com/SuperTux/supertux/issues/648

tl;dr: Arch doesn't ship with libkrb5.so.26 so it will not compile.

**Comment by probonopd** [comment]
@Peppernrino the link does not work for me. Are you saying you have a compile-time problem or does the AppImage refuse to run since libkrb5 is missing/named differently on your target system?

**Comment by tomasbrod** [comment]
https://github.com/SuperTux/supertux/issues/648 <- the link had wronk markup

**Comment by Peppernrino** [comment]
pardon that. tried to use the link button. never doing that again.

@probonopd - the library is missing. Arch only ships with libkrb5.so.3. you know this, you're in the other thread... :P

**Comment by probonopd** [comment]
So we would either have to change the code so that Kerberos is not needed, or bundle the Kerberos libraries, or bundle "fake" Kerberos libraries that do nothing, or talk Arch Linux into naming the Kerberos libraries like all other distributions do, so that the binary here would pick them up from the system. 

Which route should we take?

Also see https://github.com/probonopd/AppImages/issues/187

**Comment by tomasbrod** [comment]
Why does it need kerberos for? If it is for some crypto, why not use LibOpenSSL instead?

**Comment by caraka** [comment]
Arguing over kerberos on Arch is only a tiny part of the issue. There are 16 libraries on the AppImage exclusion list that the gridcoin wallet wants. I've been working through them, as their absense causes different issues/complaints on different OSs. I'm testing using live CDs and/or VMs of various flavours. It's easy enough to build a quality appimage that plays nice on Debian/Ubuntu, but I'm struggling after that.

**Comment by tomasbrod** [comment]
Then include that libraries in the image. If the image is to be portable, it should not depend on non-standard libraries. Can some fallback mechanism be set up? Only load the bundled lib if os does not have it installed?

**Comment by probonopd** [comment]
> There are 16 libraries on the AppImage exclusion list that the gridcoin wallet wants. I've been working through them, as their absense causes different issues/complaints on different OSs

Please open issues in https://github.com/probonopd/appimages; specifically, against https://github.com/probonopd/AppImages/blob/master/excludelist and/or https://github.com/probonopd/AppImages/blob/master/excludedeblist

The broader question, though, is why distributions still can't seem agree on the same naming schemes for even the most basic system libraries that make the core desktop Linux platform, despite LSB Desktop and all the years. So I am not too keen on supporting exotic distributions, only the mainstream ones mentioned on the AppImage website, with Arch being a borderline case since it is hard to support due to its frequent changing nature and strange deviations from what the other main ones do.

**Comment by caraka** [comment]
@tomasbrod I'm working through exactly what you suggested - getting the libraries we need included (and not excluded by default) in the AppImage. It's just going to take some time as I build and test, build and test.

@probonopd I will open issues once I've walked through the process and am not waisting your time with noob things. I'll be onto irc for a visit to talk about the exclusion list etc. I think all of these issues are resolvable. I have just been through the make file to eliminate all the unnecessary libraries still linked there that I discovered are not even used! So it's a bit of tidy up in progress.

**Comment by tomasbrod** [comment]
bump

**Comment by pwnorbitals** [comment]
Is this PR still WIP ?

**Comment by tomasbrod** [comment]
bump

**Comment by probonopd** [comment]
I don't plan any changes to it unless someone thinks a change is needed.

**Comment by sedited** [comment]
The Travis Recipe has been fundamentally changed. Can you rebase your changes?

**Comment by probonopd** [comment]
Sorry I don't have the time to do this at the moment, but feel free to ping me if you run into any issues doing it.

**Comment by jamescowens** [comment]
I am going to close this unless someone wants to pick it up.

**Comment by probonopd** [comment]
I had done the work 2 years ago and sent a pull request. Since then other unrelated changes have been made, essentially sabotaging my work. Why has this not been merged immediately when it was working? Now, 2 years later stuff has changed, and as a result you want me to invest even more time?

**Comment by jamescowens** [comment]
Actually I don’t and that is why this request was closed. In case you haven’t noticed, a different group has taken over development and I can’t speak to what happened 2 years ago. I certainly appreciate your position.

**Comment by philipswift** [comment]
I can't even remember what happened i year ago!

**Comment by probonopd** [comment]
> In case you haven’t noticed, a different group has taken over development

Sorry to hear, indeed I did not follow the project closely enough so I haven't noticed indeed. Thanks for the clarification.
